crates/taplo/src/dom/node.rs

```diff
@@ -244,24 +244,36 @@ impl Node {
         Ok(all.into_iter())
     }
 
+    /// Determine the portions of the source that correspond to this
+    /// `Node`.
+    ///
+    /// Since a symbol may appear in multiple places, a single node
+    /// may correspond to multiple `TextRange`s.
     pub fn text_ranges(&self) -> impl ExactSizeIterator<Item = TextRange> {
         let mut ranges = Vec::with_capacity(1);
 
         match self {
             Node::Table(v) => {
-                let entries = v.entries().read();
+                // Node::Table's syntax node points to a TableHeader
+                // element. If we have a parent (which should be the
+                // table element), we can use the entire text range.
+                if let Some(parent) = v.syntax().and_then(|n| n.parent()) {
+                    ranges.push(parent.text_range());
+                } else {
+                    let entries = v.entries().read();
+
+                    for (k, entry) in entries.iter() {
+                        ranges.extend(k.text_ranges());
+                        ranges.extend(entry.text_ranges());
+                    }
 
-                for (k, entry) in entries.iter() {
-                    ranges.extend(k.text_ranges());
-                    ranges.extend(entry.text_ranges());
-                }
+                    if let Some(mut r) = v.syntax().map(|s| s.text_range()) {
+                        for range in &ranges {
+                            r = r.cover(*range);
+                        }
 
-                if let Some(mut r) = v.syntax().map(|s| s.text_range()) {
-                    for range in &ranges {
-                        r = r.cover(*range);
+                        ranges.insert(0, r);
```

Comment:
This behavior seems related to what you're doing. I wonder if both can't be combined. In particular, I wonder if we should not preserve all the key and entry ranges. I have something like that in mind:

```rust
Node::Table(v) => {
    let entries = v.entries().read();
    for (k, entry) in ... { ... } // no change up to here
    let mut cover = None; // range that covers the node
    if let Some(mut r) = v.syntax().map(...) {
        for range in &ranges { ... }
        cover = Some(r); // this replaces the insert(0)
    }
    if let Some(parent) = v.syntax().and_then(|n| n.parent()) {
        cover = Some(match cover {
            None => parent.text_range(),
            Some(r) => r.cover(parent.text_range()),
        )};
    }
    if let Some(r) = cover {
        ranges.insert(0, r);
    }
}
```

Reply:
You're right; that's definitely a safer approach. I'll update.

Admittedly, I'm still a bit confused by the point of returning multiple ranges, especially when (in most cases) the first one is a superset of the rest. But the LSP implementation depends on this method, so it's probably best to keep it as is.

Reply:
> I'm still a bit confused by the point of returning multiple ranges, especially when (in most cases) the first one is a superset of the rest.

The same table can appear anywhere in a document

```toml
[foo]

[bar]
field = true

[foo.baz]
```

In this case `foo` should have 2 disjointed ranges, merging them into one would mean that the table spans the entire document, which is not true, it's especially important when trying to find out which table `field` belongs to just by looking at the cursor position.

That's the idea anyway, I might have added a range that unifies all for some reason and then pick the narrowest one, it was a while ago, the individual ranges I believe are still important.

Reply:
Thanks for the clarification, I thought it had something to do with dotted table headings and the `additional_syntaxes` stuff.

I'm not able to find the call-site where the narrowest range might be used to identify a symbol, but, I think this change shouldn't affect it. From all the testing I've done, the only difference I can find is when there's non-critical data (i.e. a comment) trailing a table. In that case, the first (largest) range includes that comment, where previously it didn't.

Reply:
I tried to play a bit with this change and indeed it seems to only change the case it's supposed to fix. In particular I wanted to see if something like that would get reordered:

```toml
[foo]
a = 1
c = 3
# b
b = 2
```

And it seems it doesn't.

However, you're current change is different from my suggestion in that it will prepend 2 new elements instead of extending the first one (which is always added when there's a parent, because if `parent()` returns `Some` then `syntax()` does too). The question being whether it's better to have a new (probably larger range) first or update the existing added range to be probably larger. I'm not sure on that one. I'm fine with both solutions:
- Merge this PR as is.
- Extend the first "artificial" range to also cover the parent if there's one.

What do you think?

Reply:
You're right, and that was an oversight. I've adjusted it to expand the (single) artificial range to cover the parent if available. Looks suspiciously like your original suggestion!